core/src/main/java/com/netflix/conductor/core/events/EventProcessor.java

```diff
@@ -1,235 +1,258 @@
-					}
-				}
-				
-				int i = 0;
-				List<Action> actions = handler.getActions();
-				for(Action action : actions) {
-					String id = msg.getId() + "_" + i++;
-					
-					EventExecution ee = new EventExecution(id, msg.getId());
-					ee.setCreated(System.currentTimeMillis());
-					ee.setEvent(handler.getEvent());
-					ee.setName(handler.getName());
-					ee.setAction(action.getAction());
-					ee.setStatus(Status.IN_PROGRESS);
-					if (executionService.addEventExecution(ee)) {
-						Future<Void> future = execute(ee, action, payload);
-						futures.add(future);
-					} else {
-						logger.warn("Duplicate delivery/execution? {}", id);
-					}
-				}
-			}
-			
-			for (Future<Void> future : futures) {
-				try {
-					future.get();
-				} catch (Exception e) {
-					logger.error(e.getMessage(), e);
-				}
-			}
-			
-			queue.ack(Arrays.asList(msg));
-			
-		} catch (Exception e) {
-			logger.error(e.getMessage(), e);
-		}
-	}
-
-	private Future<Void> execute(EventExecution ee, Action action, String payload) {
-		return executors.submit(()->{
-			try {
-				
-				logger.debug("Executing {} with payload {}", action.getAction(), payload);
-				Map<String, Object> output = actionProcessor.execute(action, payload, ee.getEvent(), ee.getMessageId());
-				if(output != null) {
-					ee.getOutput().putAll(output);
-				}
-				ee.setStatus(Status.COMPLETED);
-				executionService.updateEventExecution(ee);
-				
-				return null;
-			}catch(Exception e) {
-				logger.error(e.getMessage(), e);
-				ee.setStatus(Status.FAILED);
-				ee.getOutput().put("exception", e.getMessage());
-				return null;
-			}
-		});
-	}
-	
+    }
+
+    private void listen(ObservableQueue queue) {
+        queue.observe().subscribe((Message msg) -> handle(queue, msg));
+    }
+
+    @SuppressWarnings({"unchecked", "ToArrayCallWithZeroLengthArrayArgument"})
+    private void handle(ObservableQueue queue, Message msg) {
+        try {
+            executionService.addMessage(queue.getName(), msg);
+
+            String event = queue.getType() + ":" + queue.getName();
+            logger.debug("Evaluating message: {} for event: {}", msg.getId(), event);
+            List<EventExecution> transientFailures = executeEvent(event, msg);
+
+            if (transientFailures.isEmpty()) {
+                queue.ack(Collections.singletonList(msg));
+            } else if (queue.rePublishIfNoAck()) {
+                // re-submit this message to the queue, to be retried later
+                // This is needed for queues with no unack timeout, since messages are removed from the queue
+                queue.publish(Collections.singletonList(msg));
+            }
+        } catch (Exception e) {
+            logger.error("Error handling message: {} on queue:{}", msg, queue.getName(), e);
+        }
+    }
+
+    @SuppressWarnings("Guava")
+    @VisibleForTesting
+    Pair<EventExecution, Boolean> execute(EventExecution eventExecution, Action action, String payload) {
```

Comment:
What is the boolean going to provide, when the execution has already been retried using the `RetryUtil` and when traversing all the futures you can look up at the error and do the same.